### DISCUSSION TITLE
List breaks re-render of view with new observing

### DISCUSSION THREAD

**Discussion by jiledaa** [pointfreeco/swift-composable-architecture#2870] [Q&A] [answered]
Hi. I've migrated my app up to version 1.8 using new observing of the store. Everything seems to work fine except on iOS 16 when I add new element to the List component on another screen where the List exists and then show the list. The new element is not there, only after relaunching the app or when the list is long enough to scroll down and back up, the new added element seems to appear on top of the list correctly.

`
        List {
            ForEach(store.scope(state: \.entries, action: \.entries), id: \.state.id) { store in
                Feature(store: store)
            }
        }
`

- Adding WithPerceptionTracking inside List, neither inside ForEach, does not help. I dont have any purple warnings about perception  in runtime anyway. 
- When I replace List with ScrollView for example, it updates correctly. 
- When I'm adding new element to List on the same screen with the list, the re-rendering is also fine, it happens only when the List is not presented at time of updating. 

Has anyone similar experience?

**Comment by mbrandonw** [comment]
Hi @jiledaa, can you please reproduce the problem in a minimal project and share it? That is the only way for anyone to help.

**Reply by jiledaa** [reply]
Hi, thanks for reply. The app is quite complex and it is kind of property of some other company I'm working for. I will try to reproduce it and not to miss anything important:

```swift
import ComposableArchitecture
import SwiftUI

@main
struct ListApp: App {
    var body: some Scene {
        WindowGroup {
            FeatureView(store: Store(initialState: Feature1.State(), reducer: Feature1.init))
        }
    }
}

@Reducer
struct TextCore {
    @ObservableState
    struct State: Identifiable, Equatable {
        var id: String { text }
        var text: String
    }

    enum Action {
        case nada
    }

    var body: some ReducerOf<Self> {
        EmptyReducer()
    }
}

struct TextView: View {
    let store: StoreOf<TextCore>

    var body: some View {
        Text(store.text)
    }
}

@Reducer
struct Feature1 {
    @ObservableState
    struct State: Equatable {
        enum Tab: Equatable, Hashable {
            case feature1
            case feature2
        }

        var tab: Tab = .feature2
        var text: IdentifiedArrayOf<TextCore.State> = [.init(text: "Mazec 0"), .init(text: "Mazec 1")]
        var iterator: Int = 2
        var feature2: Feature2.State = .init()
    }

    enum Action {
        case selectTab(State.Tab)
        case add
        case feature2(Feature2.Action)
        case text(IdentifiedActionOf<TextCore>)
    }

    var body: some ReducerOf<Self> {
        Scope(state: \.feature2, action: \.feature2, child: Feature2.init)

        Reduce { state, action in
            switch action {
            case let .selectTab(tab):
                state.tab = tab
                return .none

            case .add:
                state.text.append(TextCore.State(text: "Mazec \(state.iterator)"))
                state.iterator += 1
                return .none

            case .feature2(.add):
                state.text.append(TextCore.State(text: "Mazec \(state.iterator)"))
                state.iterator += 1
                return .none

            case .text:
                return .none
            }
        }
    }
}

@Reducer
struct Feature2 {
    @ObservableState
    struct State: Equatable {
        var text: String = ""
    }

    enum Action {
        case add
    }

    var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .add:
                return .none
            }
        }
    }
}

struct Feature2View: View {
    let store: StoreOf<Feature2>

    var body: some View {
        Button(
            action: { store.send(.add) },
            label: { Text("Add Text") }
        )
    }
}

struct FeatureView: View {
    let store: StoreOf<Feature1>

    var body: some View {
        WithPerceptionTracking {
            WithViewStore(store, observe: { $0 }) { viewStore in
                TabView(selection: viewStore.binding(get: { $0.tab }, send: { Feature1.Action.selectTab($0) })) {
                    VStack {
                        List {
                            ForEach(store.scope(state: \.text, action: \.text), id: \.state.id) { store in
                                Text(store.text)
                            }
                        }
                        .background {
                            Color.black
                        }

                        Button(
                            action: { store.send(.add) },
                            label: { Text("Add Text") }
                        )
                    }
                    .tabItem { Text("Feature1") }
                    .tag(Feature1.State.Tab.feature1)

                    Feature2View(store: store.scope(state: \.feature2, action: \.feature2))
                        .tabItem { Text("Feature1") }
                        .tag(Feature1.State.Tab.feature2)
                }
            }
        }
    }
}

**Reply by jiledaa** [reply]
"Feature1" represents "MainFeature" and still uses WithViewStore, which somehow forks fine and updates view properly in this provided code. Getting rid of VithViewStore is still not completely done in my project.

**Reply by mbrandonw** [reply]
Is this code you provided actually demonstrating the problem? If so, what is the problem?

Also, it definitely is not right to mix `WithPerceptionTracking` and `WithViewStore`. And further, this code does have lots of purple warnings letting you know that something definitely is not right.

**Reply by jiledaa** [reply]
I know, I tried to reproduce the actual state of the app as much as possible. I might miss something because I'm not getting any purple warnings in my app, but still List is not updated properly. In this provided code, I'm getting warnings but updating works.

**Reply by jiledaa** [reply, marked as answer]
Even full replacement of WithViewStore from the app did not fixed the issue. I've fixed it temporarily by adding VStack inside the ForEach of the List.
